pkg/gds/admin.go

```diff
@@ -149,22 +160,224 @@ func (s *Admin) setupRoutes() (err error) {
+
+	// Refresh the double cookies for CSRF protection while using the access/refresh tokens
+	expiresAt := refreshToken.Claims.(*tokens.Claims).ExpiresAt
+	if err := admin.SetDoubleCookieTokens(c, expiresAt); err != nil {
 		c.JSON(http.StatusInternalServerError, admin.ErrorResponse("could not set cookies"))
 		return
 	}
 
-	c.JSON(http.StatusNotImplemented, admin.ErrorResponse("not implemented yet"))
+	// Return successful authentication!
+	c.JSON(http.StatusOK, out)
 }
 
+func (s *Admin) checkAuthorizedDomain(claims *idtoken.Payload) error {
+	// Fetch the claim from the idtoken payload
+	domain, ok := claims.Claims["hd"]
+	if !ok {
+		return errors.New("no hd claim to verify authorized domain with")
+	}
+
+	// Convert the domain into a string for verification
+	domains, ok := domain.(string)
+	if !ok {
+		return fmt.Errorf("claim type %T unparseable", domain)
+	}
+
+	// Search the authorized domains, if found return nil
+	for _, authorized := range s.conf.AuthorizedDomains {
+		if domains == authorized {
+			// Found an authorized domain!
+			return nil
+		}
+	}
+
+	return fmt.Errorf("%s is not in the configured authorized domains", domains)
+}
+
+// Reauthenticate allows the submission of a refresh token to reauthenticate an expired
+// or expiring access token and issues a new token pair. The access token must still be
+// provided in the Authorization header as a Bearer token, even if it is expired since
+// the access token contains the claims that need to be reissued. The refresh token is
+// posted in the request body as the credential. This method also resets the CSRF double
+// cookies to ensure that the max-age matches the duration of the refresh tokens.
 func (s *Admin) Reauthenticate(c *gin.Context) {
-	if err := admin.SetDoubleCookieTokens(c, time.Now().Add(time.Minute*10).Unix()); err != nil {
+	var (
+		err           error
+		tks           string
+		in            *admin.AuthRequest
+		out           *admin.AuthReply
+		accessClaims  *tokens.Claims
+		refreshClaims *tokens.Claims
+		accessToken   *jwt.Token
+		refreshToken  *jwt.Token
+	)
+
+	// Get the Bearer token from the Authorization header (contains access token)
+	if tks, err = admin.GetAccessToken(c); err != nil {
+		log.Warn().Err(err).Msg("reauthenticate called without access token")
+		c.JSON(http.StatusUnauthorized, admin.ErrorResponse("request is not authorized"))
+		return
+	}
+
+	// Parse the access token from the Authorization header without validating the
+	// claims, e.g. it doesn't matter if the access token is expired, but it should be
+	// signed correctly by the token server.
+	if accessClaims, err = s.tokens.Parse(tks); err != nil {
+		log.Warn().Err(err).Msg("reauthenticate called with invalid access token")
+		c.JSON(http.StatusUnauthorized, admin.ErrorResponse("request is not authorized"))
+		return
+	}
+
+	// Parse incoming JSON data from the client request (contains refresh token)
+	in = new(admin.AuthRequest)
+	if err = c.ShouldBind(&in); err != nil {
```

Comment:
What is the role of "binding" here in the context of the auth/auth flow diagram you shared?

Reply:
In the gin framework, binding is populating the `admin.AuthRequest` struct from the request; e.g. reading the raw JSON from the request body and unmarshaling. So this is what the server is doing to read the `{credential: [Google JWT credential]}` payload.